File: Vanilla/sources/zomboid/zombie/worldMap/editor/WorldMapEditorState.java
package zombie.worldMap.editor;

import java.util.ArrayList;
import se.krka.kahlua.vm.KahluaTable;
import zombie.Lua.LuaManager;
import zombie.core.Core;
import zombie.gameStates.GameState;
import zombie.gameStates.GameStateMachine;
import zombie.input.GameKeyboard;
import zombie.ui.UIManager;
import zombie.vehicles.EditVehicleState;

public final class WorldMapEditorState extends GameState {
   public static WorldMapEditorState instance;
   private EditVehicleState.LuaEnvironment m_luaEnv;
   private boolean bExit = false;
   private final ArrayList m_gameUI = new ArrayList();
   private final ArrayList m_selfUI = new ArrayList();
   private boolean m_bSuspendUI;
   private KahluaTable m_table = null;

   public void enter() {
      instance = this;
      this.load();
      if (this.m_luaEnv == null) {
         this.m_luaEnv = new EditVehicleState.LuaEnvironment(LuaManager.platform, LuaManager.converterManager, LuaManager.env);
      }

      this.saveGameUI();
      if (this.m_selfUI.size() == 0) {
         this.m_luaEnv.caller.pcall(this.m_luaEnv.thread, this.m_luaEnv.env.rawget("WorldMapEditor_InitUI"), (Object)this);
         if (this.m_table != null && this.m_table.getMetatable() != null) {
            this.m_table.getMetatable().rawset("_LUA_RELOADED_CHECK", Boolean.FALSE);
         }
      } else {
         UIManager.UI.addAll(this.m_selfUI);
         this.m_luaEnv.caller.pcall(this.m_luaEnv.thread, this.m_table.rawget("showUI"), (Object)this.m_table);
      }

      this.bExit = false;
   }

   public void yield() {
      this.restoreGameUI();
   }

   public void reenter() {
      this.saveGameUI();
   }

   public void exit() {
      this.save();
      this.restoreGameUI();
   }

   public void render() {
      byte var1 = 0;
      Core.getInstance().StartFrame(var1, true);
      this.renderScene();
      Core.getInstance().EndFrame(var1);
      Core.getInstance().RenderOffScreenBuffer();
      UIManager.useUIFBO = Core.getInstance().supportsFBO() && Core.OptionUIFBO;
      if (Core.getInstance().StartFrameUI()) {
         this.renderUI();
      }

      Core.getInstance().EndFrameUI();
   }

   public GameStateMachine.StateAction update() {
      if (!this.bExit && !GameKeyboard.isKeyPressed(65)) {
         this.updateScene();
         return GameStateMachine.StateAction.Remain;
      } else {
         return GameStateMachine.StateAction.Continue;
      }
   }

   public static WorldMapEditorState checkInstance() {
      if (instance != null) {
         if (instance.m_table != null && instance.m_table.getMetatable() != null) {
            if (instance.m_table.getMetatable().rawget("_LUA_RELOADED_CHECK") == null) {
               instance = null;
            }
         } else {
            instance = null;
         }
      }

      return instance == null ? new WorldMapEditorState() : instance;
   }

   private void saveGameUI() {
      this.m_gameUI.clear();
      this.m_gameUI.addAll(UIManager.UI);
      UIManager.UI.clear();
      this.m_bSuspendUI = UIManager.bSuspend;
      UIManager.bSuspend = false;
      UIManager.setShowPausedMessage(false);
      UIManager.defaultthread = this.m_luaEnv.thread;
   }

   private void restoreGameUI() {
      this.m_selfUI.clear();
      this.m_selfUI.addAll(UIManager.UI);
      UIManager.UI.clear();
      UIManager.UI.addAll(this.m_gameUI);
      UIManager.bSuspend = this.m_bSuspendUI;
      UIManager.setShowPausedMessage(true);
      UIManager.defaultthread = LuaManager.thread;
   }

   private void updateScene() {
   }

   private void renderScene() {
   }

   private void renderUI() {
      UIManager.render();
   }

   public void setTable(KahluaTable var1) {
      this.m_table = var1;
   }

   public Object fromLua0(String var1) {
      byte var3 = -1;
      switch(var1.hashCode()) {
      case 3127582:
         if (var1.equals("exit")) {
            var3 = 0;
         }
      default:
         switch(var3) {
         case 0:
            this.bExit = true;
            return null;
         default:
            throw new IllegalArgumentException("unhandled \"" + var1 + "\"");
         }
      }
   }

   public void save() {
   }

   public void load() {
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/WorldMapEditorState.lua
---@class WorldMapEditorState : zombie.worldMap.editor.WorldMapEditorState
---@field public instance WorldMapEditorState
---@field private m_luaEnv EditVehicleState.LuaEnvironment
---@field private bExit boolean
---@field private m_gameUI ArrayList|Unknown
---@field private m_selfUI ArrayList|Unknown
---@field private m_bSuspendUI boolean
---@field private m_table KahluaTable
WorldMapEditorState = {}

---@private
---@return void
function WorldMapEditorState:renderScene() end

---@public
---@return void
function WorldMapEditorState:yield() end

---@public
---@return WorldMapEditorState
function WorldMapEditorState:checkInstance() end

---@public
---@return void
function WorldMapEditorState:save() end

---@public
---@return GameStateMachine.StateAction
function WorldMapEditorState:update() end

---@private
---@return void
function WorldMapEditorState:saveGameUI() end

---@private
---@return void
function WorldMapEditorState:renderUI() end

---@public
---@return void
function WorldMapEditorState:reenter() end

---@public
---@return void
function WorldMapEditorState:exit() end

---@public
---@param arg0 String
---@return Object
function WorldMapEditorState:fromLua0(arg0) end

---@public
---@return void
function WorldMapEditorState:enter() end

---@public
---@param arg0 KahluaTable
---@return void
function WorldMapEditorState:setTable(arg0) end

---@private
---@return void
function WorldMapEditorState:restoreGameUI() end

---@public
---@return void
function WorldMapEditorState:render() end

---@private
---@return void
function WorldMapEditorState:updateScene() end

---@public
---@return void
function WorldMapEditorState:load() end
